### DISCUSSION TITLE
Avalonia in Browser issues in Release mode only

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#10854] [Q&A] [answered]
Hi, I tried Avalonia in browser (11 preview 6), but while it work well in Debug mode, there are lot of UI issues in Release mode. Many elements in the view dispear or not appear while using my app. 

I don't understand why it is different in Release that debug?

Can someone explain me? I know that Avalonia Browser is not production ready, but what does that mean exactly? Missing features? API breaking changes? Can I still use it for simple application? (if I avoid to use certains features)

Thanks for helping!

**Comment by maxkatz6** [comment, marked as answer]
Last time checked, if worked just fine in the release mode. There isn't much of anything planned to do with browser backend before the release. Your issue is missing lots of details what's wrong. 

First possible problem could be AOT, if it's enabled, and your project uses reflection.

**Reply by dbriard** [reply]
@maxkatz6 Thank you for your help!

Native AOT, you mean when you publish with self-contained option? No, I just hit F5 in visual studio (windows) to run the app in release mode.

Reflection, probably in custom controls, what's wrong with reflection in release mode?

My program is already too complex to share a simple repro but I will try to isolate the issue in a sample. 

**I noticed something interesting**: I have a ContentControl with DataTemplate**s**, and one of them contains severals controls in a stack panel (textblock, custom slider, custom NumberBox). Nothing appear in the ContentControl in Release... but if I remove the custom NumberBox, the other controls appear normally, and the others bugs I has in the UI (disapearing listbox, or controls) are fixed too... 

So probably something wrong with that NumberBox... I will try to look in detail. But it works normally in Debug/Release (Desktop), and in Debug (Web).

**Reply by timunie** [reply]
> Reflection, probably in custom controls, what's wrong with reflection in release mode? 

It's okay in release mode, but if you have trimming enabled, then the things that reflection tries to do may not be there as they were trimmed away. This is kind of pain in dotNET all around, not only Avalonia. You cannot trim a WPF app at all.

**Reply by dbriard** [reply]
@timunie I found the guilty :)

I had a class DecimalFormatter for the NumberBox, and in the constructor I have my WPF code below (without the try/catch) related to Culture. In the AvaloniaDiagnistics, there was a Culture Not Found warning that I haven't noticed.

May be CultureInfo "en-us" is not found on my french PC or in browser, but not sure why that work in Debug.

`public DecimalFormatter()
        {
            try
            {
                _languages = new List<CultureInfo>();
                _languages.Add(Thread.CurrentThread.CurrentCulture);
                if (Thread.CurrentThread.CurrentCulture.ToString() != "en-US")
                    _languages.Add(new CultureInfo("en-US"));
            }
            catch (Exception)
            {`
            
Anyway, thank for helping! and glad that there is not big issues with WASM release mode.

**Reply by dbriard** [reply]
I am still new to Avalonia, even if it is not Production ready, is there is critical missing feature in Browser backend that prevent to use it to start a project? Will it be "beta" when 11 will be release?

**Reply by timunie** [reply]
@dbriard that is hard to say, as it heavily depends on what you want to achieve. Most likely I'd say create a prototype and check if it works for you. I feel quite confident that it should work for 90 % of the use cases. Keep in mind that browser is sandboxed (i.e. I/O operations requires a bit of work), but that is not Avalonias fault.

**Reply by maxkatz6** [reply]
@dbriard there are critical limitations in the browser environment that developers need to keep in mind. Basically, sandboxed environment without file system access, without real threads (async/await works on a dispatcher, but no Task.Run or Thread.Start).
That's all isn't controlled by Avalonia, so we can't change any of these. From Avalonia specific features, Browser backend is done for 11.0 release. There are still some missing potential features like bitmaps or files clipboard, for example (iirc it works with a simple text only now), but these could be changed in the 11.x.

**Reply by dbriard** [reply]
@timunie @maxkatz6 Thanks to both of you! Good to know that backend is done for 11 Release, I'll investigate about the limitations.

I am using TaskScheduler a lot in the WPF app I want to port (design app with basic photo editing) so that may be an issue, I'll see if I can replace with async/await system but I am not so much familiar with them. I hope there is a wait to cancel a such task.

I am not sure what you mean by "bitmaps" in the missing features, I managed to load an image and display it, I would just need to manipulate pixels and to save it. I'll try the new file picker I've read about in the issues section.

I am not familiar with web development (except for Silverlight ^^), but I hope we can access to page URL to read parameters, read/write cookies, and more importantly that we can implement login with OAuth2 and have the token persistant. May be that is not related to Avalonia.

I also think that we can use javascript someway from c#.

Mind if I ask one last question, do you know if we can use Google Fonts? or if we have to embed the fonts as Avalonia Resources?

Thanks a lot for your help guys, so far I am really happy with what I've seen of Avalonia, I should have use it earlier!!, and the community here is really kind and helpful!

EDIT: I found answers to most of my questions in that post: https://github.com/AvaloniaUI/Avalonia/discussions/8511
and especially in that page: https://devblogs.microsoft.com/dotnet/use-net-7-from-any-javascript-app-in-net-7/

**Reply by maxkatz6** [reply]
> I am not sure what you mean by "bitmaps" in the missing features, I managed to load an image and display it, I would just need to manipulate pixels and to save it.

I meant bitmap clipboard support.

> but I hope we can access to page URL to read parameters, read/write cookies, and more importantly that we can implement login with OAuth2 and have the token persistant. May be that is not related to Avalonia.

All of this is possible with JS Interop, see https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/import-export-interop?view=aspnetcore-7.0.

**Comment by timunie** [comment]
@dbriard also consider to add a global try-catch that can write to a log file: https://docs.avaloniaui.net/docs/getting-started/unhandledexceptions

**Reply by dbriard** [reply]
Hi @timunie I have the global try catch, but the program do not crash, it just do not display correctly, and the more I use it, the more  objects disapear from the view... May be a try catch somewhere in Avalonia...

**Comment by kuiperzone** [comment]
I've been trying out the web-assembly demo/example also. The ability to run code in a browser is suddenly very interesting to me, as there is commercial interest in this.

I'm new to this  and I'm trying to orientate myself in terms of expectations.

For example, I see @maxkatz6 's comment about sandboxing:

> sandboxed environment without file system access, without real threads

OK. So my code is running in a client browser. What if I wanted to allow the user to save a file on their local system via a FilePicker? Can that be done?

What about web-sockets? Do we have those in web-assembly?

Let's say, I wanted to to write a WASM app that connected to a Kafka or Rabbitmq stream? In principle, can this be done?

And a real numpty question -- in a normal app we have "windows". What happens if I try and show a dialog or child window in WASM? What's the UI behaviour approach here?

Cheers

**Reply by dbriard** [reply]
I only can answer a few questions:

> OK. So my code is running in a client browser. What if I wanted to allow the user to save a file on their local system via a FilePicker? Can that be done?

Yes, that can be done using File picker, by writing in a stream of StorageItem object. I did not fully tried yet, but you can access to local storage of the browser, and IndexedDB. JavaScript interop work like pinvoke.

> And a real numpty question -- in a normal app we have "windows". What happens if I try and show a dialog or child window in WASM? What's the UI behaviour approach here?

For that, you can have a look at ContentDialog of FluentAvalonia project. It doesn't heritate from window, instead, it attach itself in the visual tree. I suppose because windows are only fore desktop apps.
